lib/compat/wordpress-6.2/class-gutenberg-rest-pattern-directory-controller-6-2.php

```diff
@@ -0,0 +1,140 @@
+	/**
+	 * Registers the necessary REST API routes.
+	 *
+	 * @since 5.8.0
+	 * @since 6.2.0 Added pattern directory categories endpoint.
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base . '/categories',
+			array(
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_pattern_categories' ),
+					'permission_callback' => array( $this, 'get_items_permissions_check' ),
+				)
+			),
+		);
+
+		parent::register_routes();
+	}
+
+	public function get_pattern_categories() {
+		// TODO: check about transient and if needed with locale..
+		// $transient_key = $this->get_transient_key( $query_args );
+
+		// /*
+		//  * Use network-wide transient to improve performance. The locale is the only site
+		//  * configuration that affects the response, and it's included in the transient key.
+		//  */
+		// $raw_patterns = get_site_transient( $transient_key );
+		// if ( ! $raw_patterns ) {
+		$api_url = 'http://api.wordpress.org/patterns/1.0/?categories';
+		if ( wp_http_supports( array( 'ssl' ) ) ) {
+			$api_url = set_url_scheme( $api_url, 'https' );
+		}
+
+		/*
+		* Default to a short TTL, to mitigate cache stampedes on high-traffic sites.
+		* This assumes that most errors will be short-lived, e.g., packet loss that causes the
+		* first request to fail, but a follow-up one will succeed. The value should be high
+		* enough to avoid stampedes, but low enough to not interfere with users manually
+		* re-trying a failed request.
+		*/
+		$cache_ttl      = 5;
+		$wporg_response = wp_remote_get( $api_url );
+		$raw_patterns   = json_decode( wp_remote_retrieve_body( $wporg_response ) );
+
+		if ( is_wp_error( $wporg_response ) ) {
+			$raw_patterns = $wporg_response;
+
+		} elseif ( ! is_array( $raw_patterns ) ) {
+			// HTTP request succeeded, but response data is invalid.
+			$raw_patterns = new WP_Error(
+				'pattern_api_failed',
+				sprintf(
+					/* translators: %s: Support forums URL. */
+					__( 'An unexpected error occurred. Something may be wrong with WordPress.org or this server&#8217;s configuration. If you continue to have problems, please try the <a href="%s">support forums</a>.' ),
+					__( 'https://wordpress.org/support/forums/' )
+				),
+				array(
+					'response' => wp_remote_retrieve_body( $wporg_response ),
+				)
+			);
+
+		} else {
+			// Response has valid data.
+			$cache_ttl = HOUR_IN_SECONDS;
+		}
+
+		// set_site_transient( $transient_key, $raw_patterns, $cache_ttl );
+		// }
+
+		if ( is_wp_error( $raw_patterns ) ) {
+			$raw_patterns->add_data( array( 'status' => 500 ) );
+
+			return $raw_patterns;
+		}
+
+		$response = array();
+
+		if ( $raw_patterns ) {
+			foreach ( $raw_patterns as $pattern ) {
+				$response[] = $this->prepare_response_for_collection(
+					$this->prepare_pattern_category_for_response( $pattern, $request )
```

Comment:
`$request` variable is not defined.
It causes this fatal error:
```
PHP Fatal error:  Uncaught Error: Call to a member function has_param() on null in /src/wp-includes/rest-api/endpoints/class-wp-rest-controller.php:591
```
This needs to be fixed.

Reply:
Thank you Anton!

Reply:
You are welcome, Nik.